[MRG] Update NeighborsBase 'auto' heuristic

gbolmier opened the pull request:
#### Reference Issues/PRs

Fixes #8213 

#### What does this implement/fix? Explain your changes.

Fix the current `NearestNeighbors` 'auto' heuristic which fails when the number of features is too high.

This change propose to choose tree-based methods when available and number of features doesn't exceed 15, otherwise to choose brute force.

#### Additional context

Benchmarks run and analysed [here](https://github.com/gbolmier/sklearn-neighbors-benchmark).
Read also #8213 discussion

gbolmier committed:
Update NearestNeighbors 'auto' heuristic

Change for tree-based methods when available and number of features <= 15 otherwise brute force.

Fix test using `algorithm='auto'` assuming 'auto' will select 'brute'.

Update "1.6.4.4. Choice of Nearest Neighbors Algorithm" user guide section.

Resolves: #8213

amueller commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -1458,7 +1458,7 @@ def test_k_and_radius_neighbors_duplicates():
         # Mask the first duplicates when n_duplicates > n_neighbors.
         X = np.ones((3, 1))
-        nn = neighbors.NearestNeighbors(n_neighbors=1)
+        nn = neighbors.NearestNeighbors(n_neighbors=1, algorithm='brute')
```
Why do you need this change?

gbolmier replied:
This part of the test used `algorithm='auto'`default, which selected `brute` with the original heuristic. The new heuristic selects `kd_tree` instead. As `brute` doesn't always output the same indices of duplicates than tree-based methods when `n_duplicates > n_neighbors`, the test would fail with the new heuristic.

I don't really understand why they don't all output the same indices of duplicates, but in order to preserve the original behavior of the test, I set `algorithm='brute'`

amueller commented:
Should we have some basic tests that the heuristic is actually used?

gbolmier committed:
Test auto algorithm heuristic in test_neighbors.py

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into nn_heuristic

amueller commented on doc/modules/neighbors.rst:
```diff
@@ -415,14 +415,14 @@ depends on a number of factors:
-least the same order as the number of training points, and that ``leaf_size``
-is close to its default value of ``30``.
+This choice is based on the assumption that the intrinsic dimensionality of the
+data is generally to high for tree-based methods — to provide significant
```
is this a unicode dash?

gbolmier replied:
Yep (U+2014)

rth replied:
We can't use unicode unfortunately as we still have a pdflatex build somewhere that doesn't support it.

gbolmier replied:
Alright, can replace it with a comma instead

amueller commented:
ping @rth ?

rth commented on doc/modules/neighbors.rst:
```diff
@@ -415,14 +415,14 @@ depends on a number of factors:
   will be required, brute force is better than a tree-based method.
 
-Currently, ``algorithm = 'auto'`` selects ``'brute'`` if :math:`k >= N/2`,
+Currently, ``algorithm = 'auto'`` selects ``'brute'`` if :math:`D > 15`,
```
We are sure we don't want to keep the condition `k>=N/2` in addition?

rth replied:
Yeah, maybe let's just make the explanations of it simpler.

rth reviewed (commented):
Two comments otherwise LGTM. Thanks @gbolmier !

Please add an entry to the change log at `doc/whats_new/v0.23.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself  with `:user:`.

gbolmier commented:
> Please add an entry to the change log at `doc/whats_new/v0.23.rst`.

You meant `doc/whats_new/v0.24.rst` right?

gbolmier committed:
Fix NeighborsBase 'auto' heuristic

Add the previous heuristic to the new one instead of latter replacing the former. Update test and user guide.

gbolmier committed:
Add NeighborsBase new 'auto' heuristic in what's new/v0.24.rst

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into nn_heuristic

gbolmier committed:
Fix NeighborsBase heuristic entry in changelog

rth commented on doc/modules/neighbors.rst:
```diff
@@ -415,14 +415,16 @@ depends on a number of factors:
+the number of query points is at least the same order as the number of
+training points, that ``leaf_size`` is close to its default value of ``30``,
+and that when the number of features exceeds 15, the intrinsic dimensionality
+of the data is generally to high for tree-based methods.
```
Maybe let's do bullet points for this sentence and the conditions above? As currently they are long sentences with lots of information packed that are hard to read.

"""
Currently, ``algorithm = 'auto'`` selects ``'brute'`` if any of the following conditions are verified,
- ``metric = 'precomputed'``
- or math:`D > 15`
...

Otherwise...

This heuristic is based on the following assumptions:
- the number of query points is at least
- ``leaf_size`` is close to its default value of ``30``
- when the number of features exceeds 15, ...
"""

gbolmier committed:
Update neighbors heuristic in user guide

Rephrase with bullets for readability

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into nn_heuristic

gbolmier committed:
Fix missing blank line after bullet

rth commented on doc/modules/neighbors.rst:
```diff
@@ -415,14 +415,22 @@ depends on a number of factors:
+* ``metric = 'precomputed'``
+* :math:`D > 15`
+* :math:`k >= N/2`
+* ``effective_metric_`` isn't in the ``VALID_METRICS`` list for either ``'kd_tree'`` or ``'ball_tree'``
```
Could you please wrap it to be less that 80 char long (here and below)? Otherwise looks good.

gbolmier committed:
Fix too long lines in doc/modules/neighbors.rst

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into nn_heuristic

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into nn_heuristic

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into nn_heuristic

rth reviewed (approved):
Thanks @gbolmier , LGTM.

Don't hesitate to ping on PRs in general, I though this one was already merged..
